### ISSUE TITLE
Frequency Trails: What the Mean Really Means

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#7716]
Frequency Trails: What the Mean Really Means<br>
<br>
https://ift.tt/1QWdyJm<br>
<br>
<br>
<br>
<div>
<div>
<div>
<a href="http://www.brendangregg.com/FrequencyTrails/frequencytrail_waterfall_mysqlmean35centercov_blue_z.png"><img width="250" src="http://www.brendangregg.com/FrequencyTrails/frequencytrail_waterfall_mysqlmean35centercov_blue_z_500.png" height="281"></a><br>
<center><span size="-1"><i>Mean Butterfly Plot</i></span></center>
</div>
<p>When analyzing response time, or latency, you need much more information than an average provides. The average, commonly the arithmetic mean, shows the index of central tendency. But, as I've found when studying latency distributions in production environments, the tendency is often not central, but may be skewed by <a href="http://www.brendangregg.com/FrequencyTrails/outliers.html">outliers</a>, or split by <a href="http://www.brendangregg.com/FrequencyTrails/modes.html">multiple modes</a>.</p>
<p>I previously determined often outliers and multiple modes occurred determined quantitatively, using tests and a survey of hundreds of production servers and different types of latency. I found that over 95% had six-sigma outliers, and at least 20% had multiple modes. However, nothing beats a visualization, such as a histogram, <a href="http://www.brendangregg.com/FrequencyTrails/../frequencytrails.html">frequency trail</a>, scatter plot, or <a href="http://dtrace.org/blogs/brendan/2013/05/19/revealing-hidden-latency-patterns/">heat map</a>.</p>
<p>In this post, I'll use a visualization to show what the mean really means.</p>
<a name="SummaryStatistic" id="SummaryStatistic"></a>
<h2>Summary Statistic</h2>
<p>Here's an example from the command line, from the iostat(1) tool:</p>
<pre>$ <b>iostat -xz 1</b>
[...]
Device: rrqm/s wrqm/s    r/s  w/s   rsec/s wsec/s avgrq-sz avgqu-sz <b>await</b> svctm %util
sda       0.00   0.00 256.00 0.00 64968.00   0.00   253.78     0.46 <b> 1.80</b>  1.05 27.00
[...]
</pre>
<p>Many are averages of some kind. Disk I/O response time, await, is particularly important, as applications may be blocked on this latency (newer versions split this into r_await and w_wait). The value 1.8 (ms) is the average during the interval, which leaves much to the imagination.</p>
<p>Statistics like these are also plotted as line graphs by most monitoring tools.</p>
<a name="Guessing" id="Guessing"></a>
<h2>Guessing</h2>
<p>Given an average of 1.8 ms, you may guess a distribution like this:</p>
<center><a href="http://www.brendangregg.com/FrequencyTrails/guessing_mean.png"><img src="http://www.brendangregg.com/FrequencyTrails/guessing_mean_1200.png" width="600" title="guessing_mean_1200" height="218"></a></center>
<p>This shows a <a href="http://en.wikipedia.org/wiki/Normal_distribution">normal distribution</a> centered on 1.8 ms. Maybe you guess a gamma distribution instead, or log-normal. Maybe you also have standard deviation, min and max, and 99th percentiles, and can guess more about the range of the distribution.</p>
<p>What you <em>guess</em> is important to consider, as, without additional information, that's all you can do: guess.</p>
<a name="Averages" id="Averages"></a>
<h2>Averages</h2>
<p>I previously studied <a href="http://www.brendangregg.com/FrequencyTrails/modes.html">modes</a>, visualizing them from hundreds of production servers. Those visualizations had <a href="http://www.brendangregg.com/FrequencyTrails/outliers.html">outliers</a> trimmed, to focus on the bulk of the data. Here, I'll use the same distributions to investigate averages. As the outliers have been trimmed, these averages are a type of <a href="http://en.wikipedia.org/wiki/Truncated_mean">trimmed arithmetic mean</a>, which places the mean closer to the median.</p>
<p>This is MySQL command latency from 35 random production servers. Axes and labels aren't included, as we are interested in the shape of the distribution and the average (arithmetic mean), indicated as a black line on each:</p>
<center><a href="http://www.brendangregg.com/FrequencyTrails/frequencytrail_waterfall_mysqlmean35cov_blue.png"><img src="http://www.brendangregg.com/FrequencyTrails/frequencytrail_waterfall_mysqlmean35cov_blue_1200.png" width="600" title="frequencytrail_waterfall_mysqlmean35cov_blue_1200" height="546"></a></center>
<p>Many of the averages are not where I'd guess them to be. Here is the same for <a href="http://dtrace.org/blogs/brendan/files/2013/07/frequencytrail_waterfall_mysqlmean100cov_blue.png_z.png">100 servers</a>.</p>
<p>These frequency trails are ordered top-down based on <a href="http://en.wikipedia.org/wiki/Coefficient_of_variation">CoV</a>, which locates similar shapes together.</p>
<a name="MeanButterflyPlot" id="MeanButterflyPlot"></a>
<h2>Mean Butterfly Plot</h2>
<p>Now let's see the same 35 distributions, centered on the mean:</p>
<center><a href="http://www.brendangregg.com/FrequencyTrails/frequencytrail_waterfall_mysqlmean35centercov_blue.png"><img src="http://www.brendangregg.com/FrequencyTrails/frequencytrail_waterfall_mysqlmean35centercov_blue_1200.png" width="600" title="frequencytrail_waterfall_mysqlmean35centercov_blue_1200" height="546"></a></center>
<p>Also see the <a href="http://www.brendangregg.com/FrequencyTrails/frequencytrail_waterfall_mysqlmean35centercov_white.png">white</a> and <a href="http://www.brendangregg.com/FrequencyTrails/frequencytrail_waterfall_mysqlmean35centercov_black.png">black</a> versions, and for <a href="http://www.brendangregg.com/FrequencyTrails/frequencytrail_waterfall_mysqlmean100centercov_blue_z.png">100 servers</a> (and the <a href="http://www.brendangregg.com/FrequencyTrails/frequencytrail-mean-tshirt.png">t-shirt</a> design). These look so wrong, my first thoughts was that I'd made a mistake &ndash; I went diving into my R code and datasets to see what was up. The mean is pulled more to the right than I'd expect for these distributions. I did zoom a little too much, although that <a href="http://www.brendangregg.com/FrequencyTrails/frequencytrail_waterfall_mysqlmean35zoom_blue_z.png">doesn't</a> explain it. This image itself doesn't even look centered on this page, despite using &lt;center&gt; tags.</p>
<p>Not what you expected either? Exactly. And that's a serious problem when we only have the average to go by, or the average plus other basic statistics. Our intuition is just plain wrong.</p>
<div><a href="http://www.brendangregg.com/FrequencyTrails/brendan_balance_z.jpg"><img width="200" title="brendan_balance_crop_400" src="http://www.brendangregg.com/FrequencyTrails/brendan_balance_crop_400.jpg" height="202"></a></div>
<p>I needed a physical representation to help my intuition understand this result. In the picture on the right, I'm balancing six Lego bricks on the left, with two Lego bricks on the right, in a shape similar to these distributions. Helpful?</p>
<p>This is also described in [Rice 95] when introducing the expected value of a random variable (the mean), E(X):</p>
<blockquote>E(X) is also referred to as the mean of X and is often denoted by &mu; or &mu;<sub>x</sub>. You may find it helpful to think of the expected value of X as the center of mass of the frequency function. Imagine placing the masses p(x<sub>i</sub>) at the points x<sub>i</sub> on a beam; the balance point of the beam is the expected value of X.</blockquote>
<p>As in my picture: how can six blocks balance two? In the MySQL example latency distributions, the tail is relatively long, and doesn't need as much mass to balance the closer mode on the left.</p>
<p>I've rendered the bottom distribution in the previous waterfall plot as a <a href="http://www.brendangregg.com/FrequencyTrails/mysql_hist_tall.png">tall histogram</a> with wide columns to magnify detail on the right of the mean, along with column counts to help us understand why the mean appears so offset. There are 1,371 MySQL commands to the left of the mean column, and a total of 391 to the right. For this to balance, the right hand counts would need an average distance of 3.5 columns (1371 / 391), which appears to be the case.</p>
<a name="DiskIO" id="DiskIO"></a>
<h2>Disk I/O</h2>
<p>As another example, here are 35 distributions of disk I/O latency centered on the mean:</p>
<center><a href="http://www.brendangregg.com/FrequencyTrails/frequencytrail_waterfall_diskmean35_blue.png"><img src="http://www.brendangregg.com/FrequencyTrails/frequencytrail_waterfall_diskmean35_blue_1200.png" width="600" title="frequencytrail_waterfall_diskmean35_blue_1200" height="546"></a></center>
<p>The distributions at the top look bimodal, with the average around the midpoint of the modes. These waterfall plots have the x-axis range scaled for each frequency trail. The distributions at the bottom have long tails and outliers, and a skewed average. An average alone is misleading for almost all of these distributions. (At least for the skewed averages, if the offset isn't large, it may not be a problem.)</p>
<p>Also see the <a href="http://www.brendangregg.com/FrequencyTrails/frequencytrail_waterfall_diskmean35_red_z.png">red</a> version (same color scheme as previous post for disk I/O), the <a href="http://www.brendangregg.com/FrequencyTrails/frequencytrail_waterfall_diskmean200_red_z.png">200 server</a> version, and a 35 server version which is <a href="http://www.brendangregg.com/FrequencyTrails/frequencytrail_waterfall_diskmean35full_red.png">untrimmed</a> and less centered due to the influence of outliers (but also lacks detail when extreme outliers compress the range).</p>
<a name="NextSteps" id="NextSteps"></a>
<h2>Suggested Next Steps</h2>
<p>As the distributions from these surveyed systems show, an average is not a good representation of real-world latency distributions. The best way to understand them is by visualizing them, which can be done using histograms, <a href="http://dtrace.org/blogs/brendan/2013/05/19/revealing-hidden-latency-patterns/">heat maps</a>, density plots, or <a href="http://www.brendangregg.com/FrequencyTrails/../frequencytrails.html">frequency trails</a>.</p>
<p>As a way to identify numerically whether the average has been compromised &ndash; and the distribution needs to be visualized &ndash; you can try the six-sigma test from <a href="http://www.brendangregg.com/FrequencyTrails/outliers.html">Detecting Outliers</a>, and the modal test from <a href="http://www.brendangregg.com/FrequencyTrails/modes.html">Modes and Modality</a>. These can identify whether outliers and multiple modes are present.</p>
<p>The difference between the mean and the median can be studied to detect when the mean has been offset; it may not detect bimodal distributions, where the mean and median are similar.</p>
<p>As a statistic, averages (including the arithmetic mean) have many practical uses. Properly understanding a distribution isn't one of them.</p>
<a name="Acknowledgements" id="Acknowledgements"></a>
<h2>References, Acknowledgements</h2>
<ul>
<li>Inspiration for the opening from <a href="https://twitter.com/mbostock/statuses/330003091543949313">@mbostock</a>
</li>
<li>[Rice 95] Rice, J. Mathematical Statistics and Data Analysis, 2nd Ed., Duxbury Press</li>
</ul>
<p>For more frequency trails, see the <a href="http://www.brendangregg.com/FrequencyTrails/../frequencytrails.html">main page</a>.</p>
</div>
</div>
<br>
<br>
<br>
<br>
<br>
<br>
via www.brendangregg.com https://ift.tt/RjqNDC<br>
<br>
April 3, 2021 at 05:44PM
